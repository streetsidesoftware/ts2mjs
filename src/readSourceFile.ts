import { readFile } from './fileUtils.js';
import type { SourceFile, SourceMap } from './SourceFile.js';
import { fileURLToPath, pathToFileURL } from 'url';

export const SOURCE_MAP_URL_MARKER = '//' + '# ' + 'sourceMappingURL' + '=';

export interface SourceMapRef {
    index: number;
    url: URL;
}

export async function readSourceFile(filename: string): Promise<SourceFile> {
    const content = await readFile(filename);
    const srcMapRef = getSourceMapRef(filename, content);
    const sourceMap = srcMapRef && (await readSourceMap(srcMapRef.url));

    const srcContent = (srcMapRef && content.slice(0, srcMapRef.index)) || content;

    return {
        srcFilename: filename,
        content: srcContent,
        map: sourceMap,
    };
}

export function getSourceMapRef(filename: string, content: string): SourceMapRef | undefined {
    const srcMapIdx = content.lastIndexOf(SOURCE_MAP_URL_MARKER);
    if (srcMapIdx < 0) return undefined;
    const srcMapUrl = content.slice(srcMapIdx + SOURCE_MAP_URL_MARKER.length).trim();

    try {
        const url = new URL(srcMapUrl, pathToFileURL(filename));
        return { index: srcMapIdx, url };
    } catch (e) {
        return undefined;
    }
}

export async function readSourceMap(url: URL): Promise<SourceMap | undefined> {
    try {
        return { map: await readFile(fileURLToPath(url)), url };
    } catch (e) {
        // do not break if source map file is missing.
        return undefined;
    }
}
