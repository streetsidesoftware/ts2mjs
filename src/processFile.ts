import { basename, dirname, join as pathJoin, normalize, relative, sep as pathSep } from 'path';

import assert from 'assert';
import { fileURLToPath, pathToFileURL } from 'url';
import { createMagicString, MagicString } from '../lib/magicString.mjs';
import { doesContain, isRelativePath, rebaseFile } from './fileUtils.js';
import { readSourceFile, SOURCE_MAP_URL_MARKER } from './readSourceFile.js';
import type { SourceFile, SourceMap } from './SourceFile.js';

const isSupportedFile = /\.(m?js|d\.m?ts)$/;

const regExpImportExport = /(import|export).*? from ('|")(?<file>\..*?)\2;/g;

export interface ProcessFileResult {
    filename: string;
    oldFilename: string;
    content: string;
    linesChanged?: number;
}

interface AdjustedSourceMap extends SourceMap {
    oldUrl: URL;
}

export interface ProcessSourceFileResult extends ProcessFileResult {
    mappings: AdjustedSourceMap | undefined;
}

export function processSourceFile(src: SourceFile, srcRoot: string, targetRoot: string): ProcessSourceFileResult {
    const { srcFilename, content, map: mappings } = src;

    assert(doesContain(srcRoot, srcFilename), 'Must be under root.');
    assert(isSupportedFile.test(srcFilename), 'Must be a supported file type.');

    const exp = new RegExp(regExpImportExport);

    const magicString = createMagicString(content, { filename: src.srcFilename });

    let linesChanged = 0;
    let match: RegExpExecArray | null;
    while ((match = exp.exec(content))) {
        const { index } = match;

        const line = match[0];
        const reference = match.groups?.['file'];

        if (reference && isRelativePath(reference)) {
            const start = index + line.lastIndexOf(reference);
            const end = start + reference.length;
            const newRef = rebaseImport(reference, srcFilename, srcRoot, targetRoot);
            magicString.update(start, end, newRef);
            linesChanged += 1;
            continue;
        }
    }

    const filename = calcNewFilename(srcFilename, srcRoot, targetRoot);

    const sourceMap = remapSourceMap(mappings, magicString, srcFilename, filename);

    if (!linesChanged)
        return {
            filename,
            oldFilename: srcFilename,
            content: addSourceMapToContent(content, sourceMap),
            mappings: sourceMap,
        };

    return {
        filename,
        oldFilename: srcFilename,
        content: addSourceMapToContent(magicString.toString(), sourceMap),
        linesChanged,
        mappings: sourceMap,
    };
}

export function isSupportedFileType(filename: string): boolean {
    return isSupportedFile.test(filename);
}

export async function processFile(filename: string, root: string, target: string): Promise<ProcessFileResult[]> {
    const src = await readSourceFile(filename);
    const r = processSourceFile(src, root, target);
    if (!r.mappings) return [r];

    return [r, processSourceMap(r.mappings)];
}

function calcNewFilename(srcFilename: string, root: string, target: string): string {
    const newName = srcFilename.replace(/\.js(\.map)?$/, '.mjs$1').replace(/\.ts(.map)?$/, '.mts$1');
    return rebaseFile(newName, root, target);
}

export const __testing__ = {
    calcNewFilename,
};

function processSourceMap(sourceMap: AdjustedSourceMap): ProcessFileResult {
    const pr: ProcessFileResult = {
        filename: fileURLToPath(sourceMap.url),
        oldFilename: fileURLToPath(sourceMap.oldUrl),
        content: sourceMap.map,
        linesChanged: 1,
    };

    return pr;
}

function remapSourceMap(
    sourceMap: SourceMap | undefined,
    magicString: MagicString,
    fromSourceFilename: string,
    toSourceFilename: string
): AdjustedSourceMap | undefined {
    if (!sourceMap) return undefined;

    const url = pathToFileURL(toSourceFilename + '.map');
    const map = magicString.generateMap({ file: toSourceFilename, source: fromSourceFilename });
    return { url, oldUrl: sourceMap.url, map: map.toString() };
}

function addSourceMapToContent(content: string, sourceMap: AdjustedSourceMap | undefined): string {
    if (!sourceMap) return content;

    const nl = content.endsWith('\n') ? '' : '\n';
    return content + nl + SOURCE_MAP_URL_MARKER + basename(fileURLToPath(sourceMap.url));
}

function rebaseRelFileReference(importFile: string, currentFile: string, root: string, target: string): string {
    const currentDir = dirname(currentFile);
    const targetDir = dirname(rebaseFile(currentFile, root, target));

    const importFileAbs = pathJoin(currentDir, importFile);

    const relImport = doesContain(root, importFileAbs)
        ? relative(currentDir, importFileAbs)
        : relative(targetDir, importFileAbs);
    const newImportFile = normalizeImport(relImport);
    return newImportFile;
}

function rebaseImport(importFile: string, currentFile: string, root: string, target: string): string {
    const newImportFile = rebaseRelFileReference(importFile, currentFile, root, target).replace(/\.js$/, '.mjs');
    return newImportFile;
}

function normalizeImport(relativeImport: string): string {
    relativeImport = normalize(relativeImport);
    const ref = pathSep === '\\' ? relativeImport.replace(/[\\]/g, '/') : relativeImport;
    const relRef = ('./' + ref).replace('./../', '../');
    return relRef;
}
