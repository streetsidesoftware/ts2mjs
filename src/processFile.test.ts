import { describe, expect, test } from 'vitest';

import { processSourceFile, __testing__ } from './processFile.js';
import { readSourceFile } from './readSourceFile.js';
import { resolveFixture, resolverTemp } from './test.util.js';

const ff = resolveFixture;

const oc = (obj: unknown) => expect.objectContaining(obj);

describe('processFile', () => {
    test.each`
        file                                | root                     | target      | expected
        ${'sample/lib/index.js'}            | ${'sample/lib'}          | ${'target'} | ${{ filename: 'target/index.mjs', linesChanged: 2 }}
        ${'sample/lib/index.d.ts'}          | ${'sample/lib'}          | ${'target'} | ${{ filename: 'target/index.d.mts', linesChanged: 3 }}
        ${'sample/lib/database/fetch.d.ts'} | ${'sample/lib/database'} | ${'target'} | ${{ filename: 'target/fetch.d.mts', linesChanged: 1 }}
    `('processFile $file', async ({ file, root, target, expected }) => {
        const resolveTemp = resolverTemp();
        file = ff(file);
        root = ff(root);
        target = resolveTemp(target);

        const resolvedExpected = oc({
            ...expected,
            filename: resolveTemp(expected.filename),
            oldFilename: file,
        });

        const src = await readSourceFile(file);
        const result = processSourceFile(src, root, target);
        expect(result).toEqual(resolvedExpected);
        expect(result.content).toMatchSnapshot();
    });

    test.each`
        file                          | root                | expected
        ${ff('sample/lib/image.css')} | ${ff('sample/lib')} | ${Error('Must be a supported file type.')}
        ${ff('sample/src/index.js')}  | ${ff('sample/lib')} | ${Error('Must be under root.')}
    `('processFile not handled', async ({ file, root, expected }) => {
        const content = '';
        const target = ff('../temp/lib');
        expect(() => processSourceFile({ srcFilename: file, content }, root, target)).toThrowError(expected);
    });

    test.each`
        file                               | root            | target        | expected
        ${ff('sample/lib/index.js')}       | ${ff('sample')} | ${ff('temp')} | ${ff('temp/lib/index.mjs')}
        ${ff('sample/lib/index.js.map')}   | ${ff('sample')} | ${ff('temp')} | ${ff('temp/lib/index.mjs.map')}
        ${ff('sample/lib/index.d.ts.map')} | ${ff('sample')} | ${ff('temp')} | ${ff('temp/lib/index.d.mts.map')}
        ${ff('sample/lib/style.css')}      | ${ff('sample')} | ${ff('temp')} | ${ff('temp/lib/style.css')}
    `('calcNewFilename', ({ file, root, target, expected }) => {
        expect(__testing__.calcNewFilename(file, root, target)).toEqual(expected);
    });
});
