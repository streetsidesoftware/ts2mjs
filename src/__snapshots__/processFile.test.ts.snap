// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`processFile > processFile 'sample/lib/index.d.ts' 1`] = `
"export { PrimeNumber, Tuple, GUID, AddressGuid, Address, PhoneNumberGuid, PhoneNumber, PersonGuid, Person, Annotation, } from './types.mjs';
export { lookUpPerson } from './lookup.mjs';
export { fetchEntity } from './database/index.mjs';
//# sourceMappingURL=index.d.mts.map"
`;

exports[`processFile > processFile 'sample/lib/index.js' 1`] = `
"export { lookUpPerson } from './lookup.mjs';
export { fetchEntity } from './database/index.mjs';
//# sourceMappingURL=index.mjs.map"
`;

exports[`processFile > processFile allowJsOutsideOfRoot 'sample/lib/database/fetch.d.ts' 1`] = `
"import { Entity, GUID } from '../../../../../fixtures/sample/lib/types.js';
export declare function fetchEntity<T extends Entity>(guid: GUID): Promise<T | undefined>;
//# sourceMappingURL=fetch.d.mts.map"
`;
